Add TypeScript support

ispivey opened the pull request:
The goal of this change is to eventually allow us to write snippets in TypeScript, transpile them to JavaScript, and surface both versions in the Template Gallery. By treating the TypeScript snippets as the source-of-truth, we ensure they stay in sync.

This also gives us a good way to catch potential bugs in the JavaScript versions of the snippets!

In the short term, this specific change should effectively be a no-op for the Template Gallery — the API only surfaces the JS files. If folks think this is a good idea, we can do further work to expose the TypeScript snippets in the docs. 

Eventually, we should consider removing the transpiled JavaScript snippets from source control entirely. For now, it seems more helpful to keep them in source control so reviewers can see the changes that will result from replacing the JavaScript snippets with TypeScript snippets and transpiling back to JS.

Please consider reviewing commit-by-commit, as commits are small and related in scope. 

This change does several things:
* Adds an `npm` script command to run `prettier`, which is in line with the repo standards, and uses the existing `.prettierrc`.
* Runs `prettier` and commits the small number of resulting changes
* Adds a `templates/typescript` folder, as well as a `transpile` command that uses `tsc` to transpile any TypeScript files in said folder into the `templates/javascript` folder.
* Translates all but one of the JavaScript snippets into TypeScript.
* Adds a pre-commit step to run eslint/prettier.

After translating each JS snippet into TS, I ran `npm run transpile && npm run format`. If a commit only contains a new TypeScript snippet, that means the transpiled JavaScript version was identical to the JavaScript snippet already in source control.

Eventually, I added husky/eslint as well, to ensure we're linting, transpiling, and formatting before every commit.

This change does not seek to meaningfully change or improve the snippets themselves, other than to fix bugs or typos in comments. 

Finally, note that I purposefully did not attempt to translate `signed_request.js`, because that depends on importing node's `crypto` library. We should consider rewriting that to use WebCrypto.

ispivey committed:
Adds a templates/typescript folder, and transpiles to JavaScript

As a first example, this change creates typescript/alter_headers.ts.
Then, I ran: `npm run transpile && npm run format`.

tsc overwrote javascript/alter_headers.js, and prettier formatted it
to match the repo's existing style conventions, resulting in no
actual diff to alter_headers.js -- but the code is now generated from
TypeScript!

This change also:
* Adds eslint and an .eslintrc
* Adds husky and lint-staged, which together ensure that eslint
  and prettier are run before changes are committed. This will
  help prevent style drift and make it easier for others to
  contribute.

ispivey committed:
Adds `npm run format` command to run prettier

ispivey committed:
Runs prettier on templates

ispivey committed:
Translates `aggregate_requests` snippet to TypeScript

Noteworthy changes:
* Response.Body.json() has a return value of type Any; we should explicitly
  stringify it before possibly passing it as the first argument to the
  Response constructor.
* The Response constructor expects a string as its first argument. The
  runtime does attempt to stringify arguments passed to it, but we should
  explicitly join the list of strings in `results` before passing it as an
  argument to the constructor.

ispivey committed:
Translates `ab_testing` snippet to TypeScript

ispivey committed:
Translates `auth_basic_http` snippet to TypeScript.

ispivey committed:
Translates `auth_with_headers` snippet to TypeScript.

ispivey committed:
Translates `block_on_tls_version` to TypeScript.

ispivey committed:
Translates `bulk_origin_proxies` snippet to TypeScript

ispivey committed:
Translates `bulk_redirects` snippet to TypeScript

ispivey committed:
Translated `cache_api` snippet to TypeScript

ispivey committed:
Translates `cache_ttl` snippet to TypeScript

ispivey committed:
Translates `conditional_response` snippet to TypeScript

ispivey committed:
Translates `cookie_extract` snippet to TypeScript

ispivey committed:
Translates `cors_header_proxy` snippet to TypeScript

ispivey committed:
Translates `country_code` snippet to TypeScript

ispivey committed:
Translates `debugging_tips` snippet to TypeScript

ispivey committed:
Translates `fetch_html` snippet to TypeScript

ispivey committed:
Translates `fetch_json` snippet to TypeScript

ispivey committed:
Translates `hot_link_protection` snippet to TypeScript

ispivey committed:
Translates `http2_server_push` snippet to TypeScript

ispivey committed:
Translates `modify_req_props` snippet to TypeScript

ispivey committed:
Translates `modify_res_props` snippet to TypeScript

Also, updates code to reflect that you can't pass a Response as the
`body` parameter of the Response constructor and expect the resulting
new response to inherit properties of the original Response.

ispivey committed:
Translates `post_data` snippet to TypeScript.

This change depends on updating workers-types to define the
iterable properties of FormData, like `entries()`.

ispivey committed:
Translates `post_json` snippet to TypeScript.

ispivey committed:
Translates `private_data_loss` snippet to TypeScript

ispivey committed:
Translates `redirect` snippet to TypeScript

ispivey committed:
Translates `rewrite_links_html` snippet to TypeScript

ispivey committed:
Translates `send_raw_html` snippet to TypeScript

ispivey committed:
Translates `send_raw_json` snippet to TypeScript

ispivey commented:
@koeninger @ags799 @EverlastingBugstopper too – review not required but I'm very open to any feedback or input you have.

1000hz commented on package.json:
```diff
@@ -6,11 +6,33 @@
+    "transpile": "tsc -p ./templates/typescript/tsconfig.json"
+  },
+  "devDependencies": {
+    "@typescript-eslint/eslint-plugin": "^2.33.0",
+    "@typescript-eslint/parser": "^2.33.0",
```
these got a major version bump a few days after you put in this PR. might be worth updating to the latest.

ispivey replied:
Done – only noteworthy change was setting `no-var: error` and `prefer-const: error` in the recommended ruleset.

https://github.com/typescript-eslint/typescript-eslint/pull/2001/files#diff-0259b6f21b7b64c48133978c53c66980R27-R28

1000hz commented on templates/typescript/ab_testing.ts:
```diff
@@ -0,0 +1,25 @@
+  event.respondWith(handleRequest(event.request))
+})
+
+export {}
```
what's going on with all these empty exports?

ispivey replied:
Discussed separately in chat, but the tl;dr is:
* If there's no `export` statement, TypeScript assumes all the `handleRequest` functions are being defined in the global scope, and throws an error for duplicate function definitions.
* If I do something more sensible like `export async function handleRequest...` or `export { handleRequest }` in the TypeScript version, that gets transpiled directly into the JavaScript version which we present to users – which is bad, because that causes syntax errors if you drop the snippets into the playground.

@1000hz and I discussed other options like having a tsconfig/project per snippet, but decided that was overkill and would make the experience of contributing to this repo worse, and that this is OK for now.

1000hz commented on templates/javascript/rewrite_links_html.js:
```diff
@@ -21,11 +18,9 @@ class AttributeRewriter {
 const rewriter = new HTMLRewriter()
   .on('a', new AttributeRewriter('href'))
   .on('img', new AttributeRewriter('src'))
-
```
it's unfortunate the TypeScript compiler strips out these newlines from the resulting JS output. we should consider using Babel to strip type annotations instead of `tsc`

koeninger commented on templates/javascript/conditional_response.js:
```diff
@@ -26,7 +26,7 @@ async function handleRequest(request) {
     return new Response('Block the IP 1.2.3.4', { status: 403 })
   }
   // On ASN
-  if ((request.cf || {}).asn == 64512) {
```
Was there a reason for being paranoid about the existence of request.cf in the original snippet?

EverlastingBugstopper replied:
it doesn't exist in our GCP preview workers runtime, so yes. i think we should probably keep that conditional there.

ispivey replied:
Brought it back 👍

koeninger commented on templates/typescript/cache_ttl.ts:
```diff
@@ -0,0 +1,26 @@
+async function handleRequest(request) {
```
Any reason for sometimes having an explicit return type and sometimes not?

ispivey replied:
Nope, just a little sloppy; I'll address that.

koeninger commented on templates/typescript/tsconfig.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "compilerOptions": {
```
Does this have a lot of failures if turn on strict?

ispivey replied:
It sure did! I fixed 'em and pushed a new commit.

koeninger reviewed (approved):
Seems like a good idea, glad you were inspired to do it

ispivey committed:
Enables "strict" tsc config option, and updates snippets to comply

ispivey committed:
Updates @typescript-eslint/* major version and other minor updates

As part of updating the major version from 2.X to 3.X, I cleaned up
the rule overrides in .eslintrc.js. I added `prefer-const: 0` because
the new `plugin:@typescript-eslint/recommended` ruleset sets
`prefer-const: 2`, which results in a bunch of transformations of the
ts and js templates. I'll unset that rule and re-run in a subsequent
commit.

ispivey committed:
Sets eslint prefer-const: 2 and fixes up templates to match
